zilliqa/src/consensus.rs

```diff
@@ -431,19 +437,39 @@ impl Consensus {
             return Ok(None);
         }
 
-        // Now consider whether we want to timeout - the timeout duration doubles every time, so it
-        // Should eventually have all nodes on the same view
-        let consensus_timeout_ms = self.config.consensus.consensus_timeout.as_millis() as u64;
         let head_block = self.head_block();
         let head_block_view = head_block.view();
-        let time_since_last_view_change = SystemTime::now()
-            .duration_since(self.view.last_timeout())
-            .expect("last timeout seems to be in the future...")
-            .as_millis() as u64;
-        let view_difference = self.view.get_view().saturating_sub(head_block_view);
-        let exponential_backoff_timeout = consensus_timeout_ms * 2u64.pow(view_difference as u32);
-        let next_exponential_backoff_timeout =
-            consensus_timeout_ms * 2u64.pow((view_difference + 1) as u32);
+
+        if head_block_view + 1 == self.view.get_view() && self.create_next_block_on_timeout {
+            let time_since_last_block = SystemTime::now()
+                .duration_since(self.view.last_timeout())
+                .expect("last timeout seems to be in the future...")
+                .as_millis() as u64;
+
+            let empty_block_timeout_ms =
+                self.config.consensus.empty_block_timeout.as_millis() as u64;
+
+            let transactions_count = self.transaction_pool.size();
+            if time_since_last_block > empty_block_timeout_ms || transactions_count > 0 {
+                if let Ok(Some((block, transactions))) = self.propose_new_block() {
```

Comment:
We have to extend the if-condition by
`|| time_since_last_view_change + minimum_time_left_for_empty_block >= exponential_backoff_timeout`
as you did in the `vote()` function